WebKit engine support for Badging API

beidson opened the pull request:
#### 7457143192d5ea0040dfe29917d9af8dcb49feff
<pre>
WebKit engine support for Badging API
<a href="https://bugs.webkit.org/show_bug.cgi?id=223906">https://bugs.webkit.org/show_bug.cgi?id=223906</a>
rdar://76241764

Reviewed by Tim Horton and Chris Dumez.

This patch adds engine support for setAppBadge() and setClientBadge() on Navigator objects.
<a href="https://w3c.github.io/badging/">https://w3c.github.io/badging/</a>

When Javascript updates a badge count it is sent to the embedding application via delegate callbacks.

Engine support does not mean browser support.

The standard explicitly calls out that if the User Agent does not intend to do anything with the updated
badge count then the API should not be exposed to JavaScript.
This allows for the best practice of feature detection, so web page JavaScript can put its badge count
data somewhere else that it knows will be used.

As such, WebKit support remains disabled at runtime, and only clients interested in actually using the
badge count will enable it.

This patch implements the feature by exposing a BadgeClient on Page.

The WebContent process implementation of the BadgeClient sends new badge counts to the UI process to
be forwarded along to a delegate.

For the Cocoa port:
- If the badge count update came from a WindowNavigator object, then the new badge count is sent to the
  WKUIDelegate installed on the WKWebView.
- If the badge count update came from a WorkerNavigator object (dedicated, shared, or service), then
  the new badge count is sent to the WKWebsiteDataStoreDelegate installed on the relevant data store.

Almost the entire patch is a mechanical plumbing task of moving the badge count from JS to the API client.

But an unfortunately large amount of this patch is in service of forwarding the WorkerNavigator calls from
the worker threads to the main thread and then to lookup the appropriate Page object.

C&apos;est la vie.

* Source/WTF/Scripts/Preferences/WebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/badge/BadgeClient.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/Modules/badge/EmptyBadgeClient.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/Modules/badge/Navigator+Badge.idl: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebCore/Modules/badge/NavigatorBadge.idl: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebCore/Modules/badge/WorkerBadgeProxy.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::setAppBadge):
(WebCore::Navigator::clearAppBadge):
(WebCore::Navigator::setClientBadge):
(WebCore::Navigator::clearClientBadge):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/Page.cpp:
(WebCore::m_badgeClient):
(WebCore::m_contentSecurityPolicyModeForExtension): Deleted.
* Source/WebCore/page/Page.h:
(WebCore::Page::badgeClient):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
(WebCore::WorkerNavigator::clearAppBadge):
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/page/WorkerNavigator.idl:
* Source/WebCore/workers/DedicatedWorkerThread.cpp:
(WebCore::DedicatedWorkerThread::DedicatedWorkerThread):
* Source/WebCore/workers/DedicatedWorkerThread.h:
* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WebCore::WorkerMessagingProxy::startWorkerGlobalScope):
(WebCore::WorkerMessagingProxy::postTaskToWorkerObject):
(WebCore::WorkerMessagingProxy::setAppBadge):
* Source/WebCore/workers/WorkerMessagingProxy.h:
* Source/WebCore/workers/WorkerThread.cpp:
(WebCore::WorkerThread::WorkerThread):
* Source/WebCore/workers/WorkerThread.h:
(WebCore::WorkerThread::workerBadgeProxy const):
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::ServiceWorkerThread):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::setAppBadge):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/shared/context/SharedWorkerThread.cpp:
(WebCore::SharedWorkerThread::SharedWorkerThread):
* Source/WebCore/workers/shared/context/SharedWorkerThread.h:
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
(WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy):
(WebCore::SharedWorkerThreadProxy::setAppBadge):
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::updateAppBadge):
(API::UIClient::updateClientBadge):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setAppBadgeEnabled:]):
(-[WKPreferences _appBadgeEnabled]):
(-[WKPreferences _setClientBadgeEnabled:]):
(-[WKPreferences _clientBadgeEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::updateAppBadge):
(WebKit::UIDelegate::UIClient::updateClientBadge):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setAppBadge):
(WebKit::WebProcessProxy::setClientBadge):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::workerUpdatedAppBadge):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::workerUpdatedAppBadge):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker):
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
(WebKit::WebBadgeClient::setAppBadge):
(WebKit::WebBadgeClient::setClientBadge):
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.h: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
(WebKit::WebProcess::setAppBadge):
(WebKit::WebProcess::setClientBadge):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::badgeClient):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Badging.mm: Added.
(-[BadgeDelegate updatedAppBadge:]):
(-[BadgeDelegate _webView:updatedAppBadge:]):
(-[BadgeDelegate _webView:updatedClientBadge:]):
(-[BadgeDelegate websiteDataStore:workerOrigin:updatedAppBadge:]):

Canonical link: <a href="https://commits.webkit.org/258051@main">https://commits.webkit.org/258051@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bbde66e47d301f4dbacf6623fe2c0acd6de612b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100751 "30 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110050 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170324 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10834 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/475 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93151 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107896 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8181 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34790 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22825 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77762 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91237 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3589 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24349 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87328 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1125 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3613 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/472 "Build is in progress. Recent messages:") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29204 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43848 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90212 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5526 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5390 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20184 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -195,7 +197,7 @@ void WorkerMessagingProxy::postMessageToWorkerObject(MessageWithMessagePorts&& m
 
 void WorkerMessagingProxy::postTaskToWorkerObject(Function<void(Worker&)>&& function)
-{
+{    
```
No

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -208,6 +208,9 @@ struct UIEdgeInsets;
 
 - (void)_webView:(WKWebView *)webView decidePolicyForModalContainer:(_WKModalContainerInfo *)containerInfo decisionHandler:(void (^)(_WKModalContainerDecision))decisionHandler WK_API_AVAILABLE(macos(13.0), ios(16.0));
 
+- (void)_webView:(WKWebView *)webView updatedAppBadge:(NSNumber *)badge WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
I don't think the webView updated the app badge. aren't we asking the app to update the app badge?

cdumez commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -436,4 +438,25 @@ void WorkerMessagingProxy::terminateWorkerGlobalScope()
+    if (!m_scriptExecutionContext)
+        return;
+
+    m_scriptExecutionContext->postTask([badge = WTFMove(badge), this, protectedThis = Ref { *this }] (auto&) {
```
You want to use postTaskToLoader() I believe.

beidson replied:
If I want to use "postTaskToLoader()", then "postTaskToLoader()" is terribly misnamed. 🤣

(Will do!)

cdumez commented on Source/WebCore/workers/WorkerMessagingProxy.cpp:
```diff
@@ -436,4 +438,25 @@ void WorkerMessagingProxy::terminateWorkerGlobalScope()
+        return;
+
+    m_scriptExecutionContext->postTask([badge = WTFMove(badge), this, protectedThis = Ref { *this }] (auto&) {
+        ASSERT(isMainThread());
```
If you don't use postTaskToLoader(), then this assertion won't hold for nested workers.

cdumez commented on Source/WebCore/workers/WorkerThread.h:
```diff
@@ -146,6 +149,7 @@ class WorkerThread : public WorkerOrWorkletThread {
     RefPtr<IDBClient::IDBConnectionProxy> m_idbConnectionProxy;
     RefPtr<SocketProvider> m_socketProvider;
     bool m_isInStaticScriptEvaluation { false };
+    WorkerBadgeProxy& m_workerBadgeProxy;
```
Would be nice to group it with the other proxy data members above.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2208,6 +2209,36 @@ void WebProcessProxy::getNotifications(const URL& registrationURL, const String&
     WebNotificationManagerProxy::sharedServiceWorkerManager().getNotifications(registrationURL, tag, sessionID(), WTFMove(callback));
 }
 
+void WebProcessProxy::setAppBadge(std::optional<WebPageProxyIdentifier> identifier, const SecurityOriginData& origin, std::optional<uint64_t> badge)
```
identifier should be named pageID as per the pattern in this file. identifier is too generic.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2208,6 +2209,36 @@ void WebProcessProxy::getNotifications(const URL& registrationURL, const String&
+    page->uiClient().updateAppBadge(*page, badge);
+}
+
+void WebProcessProxy::setClientBadge(std::optional<WebPageProxyIdentifier> identifier, const SecurityOriginData&, std::optional<uint64_t> badge)
```
identifier should be named pageID as per the pattern in this file. identifier is too generic.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -2208,6 +2209,36 @@ void WebProcessProxy::getNotifications(const URL& registrationURL, const String&
+void WebProcessProxy::setClientBadge(std::optional<WebPageProxyIdentifier> identifier, const SecurityOriginData&, std::optional<uint64_t> badge)
+{
+    // Client badge can only be updated from Document contexts, and therefore there has to be a Page.
+    RELEASE_ASSERT(identifier);
```
Then why are we using a std::optional for the pageID?

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -88,6 +88,8 @@
 #include "VirtualAuthenticatorManager.h"
 #endif // ENABLE(WEB_AUTHN)
 
+using namespace WebCore;
```
I thought we stopped doing this in new code because we bundle files together when building now and this often leads to conflicts?

hortont424 replied:
Ooh, yes. please don't do this!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp:
```diff
@@ -0,0 +1,51 @@
+
+void WebBadgeClient::setAppBadge(Page* page, const SecurityOriginData& origin, std::optional<uint64_t> badge)
+{
+    std::optional<WebPageProxyIdentifier> identifier;
```
pageID

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp:
```diff
@@ -0,0 +1,51 @@
+
+void WebBadgeClient::setClientBadge(Page* page, const SecurityOriginData& origin, std::optional<uint64_t> badge)
+{
+    RELEASE_ASSERT(page);
```
Then why not take in a Page&?

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Badging.mm:
```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022?

beidson committed:
WebKit engine support for Badging API
https://bugs.webkit.org/show_bug.cgi?id=223906
rdar://76241764

Reviewed by Tim Horton and Chris Dumez.

This patch adds engine support for setAppBadge() and setClientBadge() on Navigator objects.
https://w3c.github.io/badging/

When Javascript updates a badge count it is sent to the embedding application via delegate callbacks.

Engine support does not mean browser support.

The standard explicitly calls out that if the User Agent does not intend to do anything with the updated
badge count then the API should not be exposed to JavaScript.
This allows for the best practice of feature detection, so web page JavaScript can put its badge count
data somewhere else that it knows will be used.

As such, WebKit support remains disabled at runtime, and only clients interested in actually using the
badge count will enable it.

This patch implements the feature by exposing a BadgeClient on Page.

The WebContent process implementation of the BadgeClient sends new badge counts to the UI process to
be forwarded along to a delegate.

For the Cocoa port:
- If the badge count update came from a WindowNavigator object, then the new badge count is sent to the
  WKUIDelegate installed on the WKWebView.
- If the badge count update came from a WorkerNavigator object (dedicated, shared, or service), then
  the new badge count is sent to the WKWebsiteDataStoreDelegate installed on the relevant data store.

Almost the entire patch is a mechanical plumbing task of moving the badge count from JS to the API client.

But an unfortunately large amount of this patch is in service of forwarding the WorkerNavigator calls from
the worker threads to the main thread and then to lookup the appropriate Page object.

C'est la vie.

* Source/WTF/Scripts/Preferences/WebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/badge/BadgeClient.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/Modules/badge/EmptyBadgeClient.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/Modules/badge/Navigator+Badge.idl: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebCore/Modules/badge/NavigatorBadge.idl: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebCore/Modules/badge/WorkerBadgeProxy.h: Copied from Source/WebCore/page/WorkerNavigator.cpp.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::setAppBadge):
(WebCore::Navigator::clearAppBadge):
(WebCore::Navigator::setClientBadge):
(WebCore::Navigator::clearClientBadge):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/Page.cpp:
(WebCore::m_badgeClient):
(WebCore::m_contentSecurityPolicyModeForExtension): Deleted.
* Source/WebCore/page/Page.h:
(WebCore::Page::badgeClient):
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
(WebCore::WorkerNavigator::clearAppBadge):
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/page/WorkerNavigator.idl:
* Source/WebCore/workers/DedicatedWorkerThread.cpp:
(WebCore::DedicatedWorkerThread::DedicatedWorkerThread):
* Source/WebCore/workers/DedicatedWorkerThread.h:
* Source/WebCore/workers/WorkerMessagingProxy.cpp:
(WebCore::WorkerMessagingProxy::startWorkerGlobalScope):
(WebCore::WorkerMessagingProxy::postTaskToWorkerObject):
(WebCore::WorkerMessagingProxy::setAppBadge):
* Source/WebCore/workers/WorkerMessagingProxy.h:
* Source/WebCore/workers/WorkerThread.cpp:
(WebCore::WorkerThread::WorkerThread):
* Source/WebCore/workers/WorkerThread.h:
(WebCore::WorkerThread::workerBadgeProxy const):
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::ServiceWorkerThread):
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::setAppBadge):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/shared/context/SharedWorkerThread.cpp:
(WebCore::SharedWorkerThread::SharedWorkerThread):
* Source/WebCore/workers/shared/context/SharedWorkerThread.h:
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.cpp:
(WebCore::SharedWorkerThreadProxy::SharedWorkerThreadProxy):
(WebCore::SharedWorkerThreadProxy::setAppBadge):
* Source/WebCore/workers/shared/context/SharedWorkerThreadProxy.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::updateAppBadge):
(API::UIClient::updateClientBadge):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
(-[WKPreferences _setAppBadgeEnabled:]):
(-[WKPreferences _appBadgeEnabled]):
(-[WKPreferences _setClientBadgeEnabled:]):
(-[WKPreferences _clientBadgeEnabled]):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::setDelegate):
(WebKit::UIDelegate::UIClient::updateAppBadge):
(WebKit::UIDelegate::UIClient::updateClientBadge):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setAppBadge):
(WebKit::WebProcessProxy::setClientBadge):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::workerUpdatedAppBadge):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::workerUpdatedAppBadge):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker):
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.cpp: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
(WebKit::WebBadgeClient::setAppBadge):
(WebKit::WebBadgeClient::setClientBadge):
* Source/WebKit/WebProcess/WebCoreSupport/WebBadgeClient.h: Copied from Source/WebCore/workers/shared/context/SharedWorkerThread.h.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
(WebKit::WebProcess::setAppBadge):
(WebKit::WebProcess::setClientBadge):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::badgeClient):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Badging.mm: Added.
(-[BadgeDelegate updatedAppBadge:]):
(-[BadgeDelegate _webView:updatedAppBadge:]):
(-[BadgeDelegate _webView:updatedClientBadge:]):
(-[BadgeDelegate websiteDataStore:workerOrigin:updatedAppBadge:]):

Canonical link: https://commits.webkit.org/258051@main
